src/hotkey/hotkey_item.cpp

```diff
@@ -24,243 +24,347 @@
-			(event.keysym.mod & KMOD_META)  != 0,
-			(event.keysym.mod & KMOD_ALT)   != 0
-			);
+	if (cfg["alt"].to_bool())
+		mods |=  KMOD_ALT;
+
+
+	base->set_mods(mods);
+	base->set_command(cfg["command"].str());
+
+	return base;
 }
 
+bool hotkey_mouse::matches_helper(const SDL_Event &event) const
+{
+	if (event.type != SDL_MOUSEBUTTONUP &&
+			event.type != SDL_MOUSEBUTTONDOWN)
+		return false;
 
+	if (event.button.button != button_)
+		return false;
 
-std::string get_names(std::string id) {
+	return true;
+}
 
-	std::vector<std::string> names;
-	BOOST_FOREACH(const hotkey::hotkey_item& item, hotkeys_) {
-		if (item.get_command() == id && (!item.null()) ) {
-			names.push_back(item.get_name());
-		}
-	}
+const std::string hotkey_mouse::get_name_helper() const
+{
+	return "mouse " + boost::lexical_cast<std::string>(button_);
+}
 
-	return boost::algorithm::join(names, ", ");
+void hotkey_mouse::save_helper(config &item) const
+{
+	item["mouse"] = 0;
+	if (button_ != 0) item["button"] = button_;
 }
 
-const hotkey_item& get_hotkey(int mouse, int joystick, int button, int hat, int value,
-		bool shift, bool ctrl, bool cmd, bool alt)
+const std::string hotkey_keyboard::get_name_helper() const
 {
-	return ::get_hotkey(mouse, joystick, button, hat, value,
-			shift, ctrl, cmd, alt);
+	std::string ret = std::string(SDL_GetKeyName(keycode_));
+	if (ret.size() == 1)
+		boost::algorithm::to_lower(ret);
+	return ret;
 }
 
-const hotkey::hotkey_item& get_hotkey(int character, int keycode,
-		bool shift, bool ctrl,	bool cmd, bool alt)
+bool hotkey_keyboard::matches_helper(const SDL_Event &event) const
 {
-	return ::get_hotkey(character, keycode,
-			shift, ctrl, cmd, alt);
+	if (event.type != SDL_KEYDOWN &&
+			event.type != SDL_KEYUP)
+		return false;
+
+	if (event.key.keysym.sym != keycode_)
+		return false;
+
+	return true;
+}
+
+bool hotkey_mouse::bindings_equal_helper(hotkey_ptr other) const
+{
+	hotkey_mouse_ptr other_m = boost::dynamic_pointer_cast<hotkey_mouse>(other);
+	if (other_m == hotkey_mouse_ptr())
+		return false;
+	return button_ == other_m->button_;
+}
+
+void hotkey_keyboard::save_helper(config &item) const
+{
+	if (keycode_ != SDLK_UNKNOWN) item["key"] = SDL_GetKeyName(keycode_);
 }
 
 bool has_hotkey_item(const std::string& command)
 {
-	BOOST_FOREACH(hotkey_item& item, hotkeys_)
+	BOOST_FOREACH(hotkey_ptr item, hotkeys_)
 	{
-		if(item.get_command() == command)
+		if(item->get_command() == command)
 			return true;
 	}
 	return false;
 }
 
-void add_hotkey(const hotkey_item& item) {
+bool hotkey_keyboard::bindings_equal_helper(hotkey_ptr other) const
+{
+	hotkey_keyboard_ptr other_k = boost::dynamic_pointer_cast<hotkey_keyboard>(other);
+	if (other_k == hotkey_keyboard_ptr())
+		return false;
+
+	return keycode_ == other_k->keycode_;
+}
+
+void del_hotkey(hotkey_ptr item) {
+	if (hotkeys_.size() > 0)
+		hotkeys_.erase(std::remove(hotkeys_.begin(), hotkeys_.end(), item));
+}
+
+void add_hotkey(const hotkey_ptr item) {
```

Comment:
null check is done in the bindings_equal_helper callee.

Not needed in del_hotkey, it'd simply try to delete a null element.